p2p/tests/node.rs

```diff
@@ -0,0 +1,220 @@
+    task: JoinHandle<()>,
+    /// The RPC client
+    /// Used to communicate with the p2p node.
+    client: P2pClient,
+    /// The node's peer_id
+    peer_id: PeerId,
+    /// A channel to read the network events received by the node.
+    network_events: Receiver<NetworkEvent>,
+    /// The listening address for this node.
+    addr: Multiaddr,
+}
+
+impl Drop for TestRunner {
+    fn drop(&mut self) {
+        self.task.abort();
+    }
+}
+
+#[tokio::test]
+async fn test_local_peer_id() -> Result<()> {
+    let test_runner = TestRunnerBuilder::new().build().await?;
+    let got_peer_id = test_runner.client.local_peer_id().await?;
+    assert_eq!(test_runner.peer_id, got_peer_id);
+    Ok(())
+}
+
+#[test(tokio::test(flavor = "multi_thread", worker_threads = 2))]
+async fn test_two_nodes() -> Result<()> {
+    let test_runner_a = TestRunnerBuilder::new().build().await?;
+    let test_runner_b = TestRunnerBuilder::new().build().await?;
+    let addrs_b = vec![test_runner_b.addr.clone()];
+    debug!(?test_runner_a.peer_id, ?test_runner_b.peer_id, "peer ids");
+
+    let peer_id_a = test_runner_a.client.local_peer_id().await?;
+    assert_eq!(test_runner_a.peer_id, peer_id_a);
+    let peer_id_b = test_runner_b.client.local_peer_id().await?;
+    assert_eq!(test_runner_b.peer_id, peer_id_b);
+
+    // connect
+    test_runner_a.client.connect(peer_id_b, addrs_b).await?;
+
+    // Make sure the peers have had time to negotiate protocols
+    tokio::time::sleep(Duration::from_millis(3_000)).await;
+
+    is_bi_connected(&test_runner_a, &test_runner_b).await?;
+
+    Ok(())
+}
+
+#[test(tokio::test(flavor = "multi_thread", worker_threads = 2))]
+async fn test_three_nodes() -> Result<()> {
+    let test_runner_a = TestRunnerBuilder::new().build().await?;
+    let test_runner_b = TestRunnerBuilder::new().build().await?;
+    let test_runner_c = TestRunnerBuilder::new().build().await?;
+
+    let peer_id_a = test_runner_a.client.local_peer_id().await?;
+    assert_eq!(test_runner_a.peer_id, peer_id_a);
+    let peer_id_b = test_runner_b.client.local_peer_id().await?;
+    assert_eq!(test_runner_b.peer_id, peer_id_b);
+    let peer_id_c = test_runner_c.client.local_peer_id().await?;
+    assert_eq!(test_runner_c.peer_id, peer_id_c);
+
+    // connect a to b
+    test_runner_a
+        .client
+        .connect(peer_id_b, vec![test_runner_b.addr.clone()])
+        .await?;
+    // connect b to c
+    test_runner_b
+        .client
+        .connect(peer_id_c, vec![test_runner_c.addr.clone()])
+        .await?;
+
+    // Make sure the peers have had time to negotiate protocols
+    tokio::time::sleep(Duration::from_millis(3_000)).await;
+
+    assert!(is_bi_connected(&test_runner_a, &test_runner_b).await?);
+    assert!(is_bi_connected(&test_runner_b, &test_runner_c).await?);
+    // We expect that a and c find each other through b and become connected
+    assert!(is_bi_connected(&test_runner_a, &test_runner_c).await?);
```

Comment:
This is the key test, it shows peers can find each other transitively and connect.